### PR TITLE
upgrade rustdoc's pulldown-cmark to 0.4.1

### LINKED ISSUE #59194: Update to last pulldown-cmark version
Last version (0.3.0) is finally here!

### PR DESCRIPTION
**Pull request by euclio**
Fixes #59194.

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Thanks for doing it! Please ping me once CI passed so I can re-review it (but currently looks good).

**Comment by euclio** [comment]
@GuillaumeGomez Travis is green!

File: src/librustdoc/markdown.rs [hunk 1]
```diff
@@ -143,10 +143,9 @@ pub fn test(mut options: Options, diag: &errors::Handler) -> i32 {
                                        options.linker, options.edition, options.persist_doctests);
     collector.set_position(DUMMY_SP);
     let codes = ErrorCodes::from(UnstableFeatures::from_environment().is_nightly_build());
-    let res = find_testable_code(&input_str, &mut collector, codes);
-    if let Err(err) = res {
-        diag.span_warn(DUMMY_SP, &err.to_string());
-    }
+
+    find_testable_code(&input_str, &mut collector, codes);
```
Comment (GuillaumeGomez) [inline src/librustdoc/markdown.rs]: Why those changes?

File: src/librustdoc/passes/mod.rs [hunk 2]
```diff
@@ -333,24 +333,24 @@ pub fn look_for_tests<'tcx>(
...
-                  tests.found_tests > 0 &&
-                  !cx.renderinfo.borrow().access_levels.is_doc_reachable(item.def_id) {
-            let mut diag = cx.tcx.struct_span_lint_hir(
-                lint::builtin::PRIVATE_DOC_TESTS,
-                hir::CRATE_HIR_ID,
-                span_of_attrs(&item.attrs),
-                "Documentation test in private item");
-            diag.emit();
-        }
+    find_testable_code(&dox, &mut tests, ErrorCodes::No);
```
Comment (GuillaumeGomez) [inline src/librustdoc/passes/mod.rs]: Same question.

File: src/librustdoc/test.rs [hunk 3]
```diff
@@ -796,11 +796,7 @@ impl<'a, 'hir> HirCollector<'a, 'hir> {
         // anything else, this will combine them for us.
         if let Some(doc) = attrs.collapsed_doc_value() {
             self.collector.set_position(attrs.span.unwrap_or(DUMMY_SP));
-            let res = markdown::find_testable_code(&doc, self.collector, self.codes);
-            if let Err(err) = res {
-                self.sess.diagnostic().span_warn(attrs.span.unwrap_or(DUMMY_SP),
-                    &err.to_string());
-            }
+            markdown::find_testable_code(&doc, self.collector, self.codes);
```
Comment (GuillaumeGomez) [inline src/librustdoc/test.rs]: Same question.

File: src/librustdoc/markdown.rs (hunk 1 above)
Comment (euclio) [reply src/librustdoc/markdown.rs]: While I was making the changes to `find_testable_code`, I noticed that the error case isn't actually possible. The code attempts to handle a missing codeblock end event but I don't think `pulldown-cmark` can actually emit that sequence of events.  Furthermore, there isn't a test case in the compiler that emits the error message.

So, I removed the error case, changed the signature from `Result` to `()`, and updated all the call-sites to remove the error emitting code.

Comment (GuillaumeGomez) [reply src/librustdoc/markdown.rs]: I see, good catch then!

**Comment by euclio** [comment]
Rebased.

File: src/librustdoc/markdown.rs (hunk 1 above)
Comment (Phrohdoh) [reply src/librustdoc/markdown.rs]: https://github.com/rust-lang/rust/issues/60319

**Pull request merged**